src/System.Runtime.Analyzers/Core/CallGCSuppressFinalizeCorrectly.cs

```diff
@@ -27,42 +30,221 @@ public abstract class CallGCSuppressFinalizeCorrectlyAnalyzer : DiagnosticAnalyz
+                    }
+
+                    // Checks for GC.SuppressFinalize(this)
+                    if (_invocationExpression.ArgumentsInSourceOrder.Count() != 1)
+                    {
+                        return;
+                    }
+
+                    var parameterSymbol = _semanticModel.GetSymbolInfo(_invocationExpression.ArgumentsInSourceOrder.Single().Syntax).Symbol as IParameterSymbol;
+                    if (parameterSymbol == null || !parameterSymbol.IsThis)
+                    {
+                        context.ReportDiagnostic(_invocationExpression.Syntax.CreateDiagnostic(
+                            NotPassedThisRule,
+                            _containingMethodSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat),
+                            _gcSuppressFinalizeMethodSymbol.ToDisplayString(SymbolDisplayFormats.ShortSymbolDisplayFormat)));
+                    }
+                }
+
+            }
+
+            private SuppressFinalizeUsage GetAllowedSuppressFinalizeUsage(IMethodSymbol method)
+            {
+                // We allow constructors in sealed types to call GC.SuppressFinalize.
+                // This allows types that derive from Component (such SqlConnection)
+                // to prevent the finalizer they inherit from Component from ever 
+                // being called.
+                if (method.ContainingType.IsSealed && method.IsConstructor() && !method.IsStatic)
+                {
+                    return SuppressFinalizeUsage.CanCall;
+                }
+
+                if (!IsDisposeMethod(method))
+                {
+                    return SuppressFinalizeUsage.MustNotCall;
+                }
+
+                // If the Dispose method is declared in a sealed type, we do
+                // not require that the method calls GC.SuppressFinalize
+                var hasFinalizer = method.ContainingType.HasFinalizer();
+                if (method.ContainingType.IsSealed && !hasFinalizer)
+                {
+                    return SuppressFinalizeUsage.CanCall;
+                }
+
+                // We don't require that non-public types call GC.SuppressFinalize 
+                // if they don't have a finalizer as the owner of the assembly can 
+                // control whether any finalizable types derive from them.
+                if (method.ContainingType.DeclaredAccessibility != Accessibility.Public && !hasFinalizer)
+                {
+                    return SuppressFinalizeUsage.CanCall;
+                }
+
+                // Even if the Dispose method is declared on a type without a
+                // finalizer, we still require it to call GC.SuppressFinalize to
+                // prevent derived finalizable types from having to reimplement
+                // IDisposable.Dispose just to call it.
+                return SuppressFinalizeUsage.MustCall;
+            }
+
+            private bool IsDisposeMethod(IMethodSymbol method)
```

Comment:
Seems like a good candidate to move into AnalyzerUtilities.IMethodSymbolExtensions?